### ISSUE TITLE
Backend Error: Generics, Callable, Nothing satisfies Anything, missing Cast?

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon-compiler#2002]
This:

```ceylon
[<Head|Tail>*] seqBuilder
        <Head, Tail>
        (Head head, [Tail*]() tail)()
        given Tail satisfies Anything
    => tail().withLeading(head);

[<String|Integer>*]() a1Builder =
        seqBuilder("A", // error
        seqBuilder(1,
        () => []));
```

produces:

```
method seqBuilder in class sandbox.ceylon.snap.base.seqBuilder_ cannot be applied to given types;
  required: com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor,com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor,  Head,                   ceylon.language.Callable<             ? extends ceylon.language.Sequential<? extends Tail>>
  found:    com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor,com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor,  ceylon.language.String, ceylon.language.Callable<capture#1 of ? extends ceylon.language.Sequential<? extends java.lang.Object>>
  reason: actual argument ceylon.language.Callable<capture#1 of ? extends ceylon.language.Sequential<? extends java.lang.Object>> cannot be converted to
                          ceylon.language.Callable<             ? extends ceylon.language.Sequential<? extends ceylon.language.Integer>> by method invocation conversion 
```

The generated code for `seqBuilder` does indeed return a callable with `…<? extends java.lang.Object>>`, yet `seqBuilder` expects the value for `tail` (also produced by a call to `seqBuilder` in this case) to have precise type information (i.e. `…<? extends ceylon.language.Integer>>`).

If `given Tail satisfies Anything` is commented out, the generated code includes a cast for the result of the inner call to `seqBuilder` with the necessary type information (the result is narrowed to `...? extends Integer...`).

---------------

Now, with a narrower constraint on `Tail`, the problem appears earlier, and perhaps more clearly:

```ceylon
[<Head|Tail>*] seqBuilder
        <Head, Tail>
        (Head head, [Tail*]() tail)()
        given Tail satisfies Boolean
    => tail().withLeading(head);

[<Boolean>*]() a1Builder =
        seqBuilder(true, // error
        () => []);
```

In this case, the callable `tail` is expected to produce a `Sequence` of `Boolean`s, which of course is not what Java thinks `() => []` does.

Removing the constraint on `Tail` causes the generated code to expect *just* a `Sequence` of `Object`s, which works, because that is what `() => []` produces.

-----------------------

Finally, removing the constraint on `Tail`, but providing explicit types when calling `seqBuilder`, as in:

```ceylon
[<Boolean>*]() a1Builder =
        seqBuilder<Boolean, Boolean>(true,
        () => []);
```

also works. In this case, the generated code includes the necessary cast on `() => []`:

```
(.ceylon.language.Callable<? extends .ceylon.language.Sequential<? extends .ceylon.language.Boolean>>)(.ceylon.language.Callable) …
```

**Comment by gavinking** [comment]
Confirmed. Note that this code works perfectly on js. ;-)

**Comment by quintesse** [comment]
The only difference between the two cases is that `applyErasureAndBoxing()` is called with the extra flag `EXPR_EXPECTED_TYPE_HAS_CONSTRAINED_TYPE_PARAMETERS` in the case that fails. Because no casts get added at all. I can make it work manually by forcing a cast (either a proper one or a raw cast, both work). But I have no idea what combination of checks would fix this.

**Comment by quintesse** [comment]
Hopefully the guru of erasure and boxing has an idea, @FroMage ?

**Comment by FroMage** [comment]
I'd have to understand what this is about, which I don't ATM.

**Comment by tombentley** [comment]
The test now passes, so the underlying causes has presumably been fixed.

**Comment by jvasileff** [comment]
I can run the examples now, but the generated code is the same. It is still missing casts, and I guess compiles due to #2048. Is this expected/desired? If so, would it make sense to eliminate the casts that *are* added, to simplify the code?

In the first example above, the return type of `seqBuilder` is:

```java
ceylon.language.Callable<? extends .ceylon.language.Sequential<? extends java.lang.Object>>
```

which we are passing as an argument to the `tail` parameter, which expects:

```java
ceylon.language.Callable<? extends .ceylon.language.Sequential<? extends Tail>>
```

In the failure case, this is:

```java
ceylon.language.Callable<? extends .ceylon.language.Sequential<? extends Integer>>
```

which resulted the original `javac` error.

**Comment by tombentley** [comment]
Yes, I guess #2048 is the reason this now works.

> Is this expected/desired? 

Well that's quite an interesting question. There have been a lot of compiler bugs where problem has been getting the java type system (and specifically generic types) to work right with Ceylon types. Usually these errors come to light when assigning an expression of one type to something of another type. The solution has been to add more typecasts in the relevant spots, so much so, the compiler will often do multiple typecasts, such as a raw cast then a generic cast in order to get javac to see a generic type it finds acceptable. Those multiple typecasts are pointless in a sense, because ultimately the JVM sees the erased type there anyway, so doing more than one is a noop. The only thing they were doing for us was providing a safety net that the code generator was not completely broken: javac would tell us (well, our users) when the types were inconsistent. The problem was that being able to reason about java's type system, has over time, added a lot of complexity to the code generator.

In other contexts (such as inheritence) it's less obviously a good thing that we're removing this safety net, as it *might* affect the declarations javac generates (generation of bridge methods for example), but we've not encountered any problems there yet.

> If so, would it make sense to eliminate the casts that are added, to simplify the code?

There are places where we could now remove typecasts from the generated code. But I think it would make sense to be conservative about doing that until we're use the fix to #2048 doesn't have any unintended consequences.
